Password optional Sign up Javascript Node Guide (#3435)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/guides/pwd-optional-new-sign-up-email/main/nodeexpress/integrationsteps.md

### 1: Add sign up link on sign-in page

> Add a **Sign up** link on your app's sign-in page. When the user clicks this link, they are redirected to a sign-up page where they can sign up for a new account.

Author: updated.

> <div class="common-image-format bordered-image">

Author: updated.

![Screenshot showing a sign in page with a sign up link.](/img/pwd-optional/pwd-optional-sign-up-link-sign-in-page.png)

</div>

### 2: Create and display sign-up page

Create a sign-up page that captures the user's first name, last name, and email.

> <div class="common-image-format bordered-image">

Author: updated.

> ![Screenshot showing a sign-up page with first name, last name, and email fields.](/img/pwd-optional/pwd-optional-sign-up-page.png)

Author: updated.

```json
{
  firstName: "John",
  lastName: "Doe",
  email: "johndoe@email.com",
}
```

Send this new object to `OktaAuth.idx.register()`.

```javascript
  const authClient = getAuthClient(req);
  const transaction = await authClient.idx.register(req.body);
  handleTransaction({ req, res, next, authClient, transaction });
```

### 4. Verify identity with the email authenticator

`OktaAuth.idx.register()` returns an `IdxTransaction` object indicating that the user needs to verify their identity with the email authenticator challenge.

```json
{
  status: "PENDING",
  nextStep: {
    name: "enroll-authenticator",
    type: "email",
    authenticator: {
      type: "email",
      key: "okta_email",
      displayName: "Email",
    },
  },
}
```

The email authenticator supports OTP and magic links, and you can integrate both methods into your application. Learn more about integrating the email authenticator challenge by visiting the [Okta email integration guide](/docs/guides/authenticators-okta-email/nodeexpress/main/#integrate-email-challenge-with-magic-links).

### 5. Display remaining optional authenticators

> After the user verifies their identity using the email authenticator, `OktaAuth.idx.proceed()` returns an `IdxTransaction` object indicating that the user can enroll in additional optional authenticators including the password authenticator. The `IdxTransaction.nextStep.canSkip` property identifies whether or not the remaining authenticators are optional. In this case, the property is set to `true`.

Author: updated.

```json
{
  status: "PENDING",
  nextStep: {
    name: "select-authenticator-enroll",
    options: [
      {
        label: "Okta Verify",
        value: "okta_verify",
      },
      {
        label: "Password",
        value: "okta_password",
      },
    ],
    canSkip: true,
  },
}
```

> Create and display a page that lists the remaining optional authenticators and allows the user to skip registering any additional authenticators.

Author: The new sentence isn't rolling off my tongue as easily as what I had. We can sync up about this one.

> Create and display a page that lists the remaining optional authenticators and allows the user to skip registering any additional authenticators.

Author: Add a new sentence to clarify when there are required. If there are additional required authenticators the skip button won't be there.

> <div class="common-image-format bordered-image">

Author: Is half and border in master yet? I'm not seeing the formattting.

> <div class="common-image-format bordered-image">

Reviewer: The styles are there - Chris merged them last week.
Find them in packages\@okta\vuepress-theme-prose\assets\css\okta\global\image-sizing.scss

> <div class="common-image-format bordered-image">

Author: After I merge in this PR, I'll pull in the lastest from the master branch to get the border style.

> ![Screenshot showing a page to skip the remaining optional authenticators.](/img/pwd-optional/pwd-optional-sign-up-authenticators-page.png)

Author: updated.

</div>

### 6. Skip remaining optional authenticators and complete sign-up

> When the user opts to skip registering any more authenticators, call `OktaAuth.idx.proceed()` passing in an object with a `skip` property equal to `true`.

Author: updated.

```javascript
  const authClient = getAuthClient(req);
  const transaction = await authClient.idx.proceed({ skip: true });
  handleTransaction({ req, res, next, authClient, transaction });
```

> `OktaAuth.idx.proceed()` should return  `IdxTransaction.status` of `SUCCESS` along with access and ID tokens, which indicates a successful new user sign-up.

Reviewer: ```suggestion
`OktaAuth.idx.proceed()` should return an `IdxTransaction.status` of `SUCCESS` along with access and ID tokens, which indicates a successful new user sign-up.
```

Author: updated.

> ### 1: Add sign up link on sign-in page

Reviewer 2: ```suggestion
### 1: Add sign-up link on sign-in page
```

Add a **Sign up** link on your app's sign-in page. When the user clicks this link, redirect them to a sign-up page where they can sign up for a new account.

<div class="half border">

> ![Screenshot showing a sign in page with a sign up link.](/img/pwd-optional/pwd-optional-sign-up-link-sign-in-page.png)

Reviewer 2: ```suggestion
![Screenshot showing a sign-in page with a sign-up link.](/img/pwd-optional/pwd-optional-sign-up-link-sign-in-page.png)
```

Author: updated.

</div>

### 2: Create and display sign-up page

Create a sign-up page that captures the user's first name, last name, and email.

<div class="half border">

![A sign-up page with first name, last name, and email fields, and submit button.](/img/pwd-optional/pwd-optional-sign-up-page.png)

</div>

### 3: Submit new user account details

When the user submits their account details, create an object with `firstName`, `lastName`, and `email` properties and assign them the values entered by the user.

> > **Note**: that the `email` property represents the account's username and primary email address.

Reviewer 2: ```suggestion
> **Note**: The `email` property represents the account's username and primary email address.
```

Author: updated.

    },
  },
}
```

The email authenticator supports OTP and magic links, and you can integrate both methods into your application. Learn more about integrating the email authenticator challenge by visiting the [Okta email integration guide](/docs/guides/authenticators-okta-email/nodeexpress/main/#integrate-email-challenge-with-magic-links).

### 5. Display remaining optional authenticators

After the user verifies their identity using the email authenticator, `OktaAuth.idx.proceed()` returns an `IdxTransaction` object indicating that the user has the option to enroll in additional authenticators. The `IdxTransaction.nextStep.canSkip` property is `true` if the remaining authenticators are optional.

```json
{
  status: "PENDING",
  nextStep: {
    name: "select-authenticator-enroll",
    options: [
      {
        label: "Okta Verify",
        value: "okta_verify",
      },
      {
        label: "Password",
        value: "okta_password",
      },
    ],
    canSkip: true,
  },
}
```

Create and display a page that lists the remaining optional authenticators and allows the user to skip registering any additional authenticators.

<div class="half border">

> ![A form showing a list of authenticators to enroll, and submit, cancel, and skip buttons.](/img/pwd-optional/pwd-optional-sign-up-authenticators-page.png)

Reviewer 2: ```suggestion
![A form showing a list of authenticators to enroll, and Submit, Cancel, and Skip buttons.](/img/pwd-optional/pwd-optional-sign-up-authenticators-page.png)
```

Author: updated.

</div>

> >**Note:** In other use cases where there are additional required authenticators, `IdxTransaction.nextStep.canSkip` equals `false` and the skip button should be omiited.

Reviewer 2: ```suggestion
>**Note:** In other use cases where there are additional required authenticators, `IdxTransaction.nextStep.canSkip` equals `false` and the **Skip** button should be omitted.
```

Author: updated.

> ### 6. Skip remaining optional authenticators and complete sign-up

Reviewer 2: ```suggestion
### 6. Skip remaining optional authenticators and complete sign-up flow
```

Author: updated.

> When the user clicks the **Skip** button, call `OktaAuth.idx.proceed()` passing in an object with a `skip` property equal to `true`.

Reviewer 2: ```suggestion
When the user clicks the **Skip** button, call `OktaAuth.idx.proceed()` and pass in an object with a `skip` property equal to `true`.
```

```javascript
  const authClient = getAuthClient(req);
  const transaction = await authClient.idx.proceed({ skip: true });
  handleTransaction({ req, res, next, authClient, transaction });
```

> `OktaAuth.idx.proceed()` should return an `IdxTransaction.status` of `SUCCESS` along with access and ID tokens, which indicates a successful new user sign-up.

Reviewer 2: ```suggestion
`OktaAuth.idx.proceed()` should return an `IdxTransaction.status` of `SUCCESS` along with access and ID tokens, which indicates a successful new user sign-up flow.
```

Author: updated.
